Fix load path ordering

nojb opened the pull request:
This PR fixes some bugs where the entries in the load path are passed/searched in the wrong order (reversed) in a few places. To keep in mind when reading the code: earlier entries in the load path take precedence over later entries (ie written later in the command line).

### Places where the entries were being used/recorded in the wrong order:

- In the slow path of the load path lookup: https://github.com/ocaml/ocaml/blob/e68b75eb823c27894c5a2fcf30f1166029abffb1/utils/load_path.ml#L83 and https://github.com/ocaml/ocaml/blob/e68b75eb823c27894c5a2fcf30f1166029abffb1/utils/load_path.ml#L89
(`Misc.find_in_path*` functions expect the list of directories in the normal order but `Load_path.get_paths ()` returns it reversed).
- When using `#remove_directory` in the toplevel, the load path is refreshed in the wrong order:
https://github.com/ocaml/ocaml/blob/e68b75eb823c27894c5a2fcf30f1166029abffb1/utils/load_path.ml#L65-L68
(because `!dirs` is kept in reverse, and so the `List.iter` call is giving precedence to the later entries in the load path, instead of earlier ones).
- When passed to the linker: https://github.com/ocaml/ocaml/blob/e68b75eb823c27894c5a2fcf30f1166029abffb1/utils/ccomp.ml#L206 and https://github.com/ocaml/ocaml/blob/e68b75eb823c27894c5a2fcf30f1166029abffb1/utils/ccomp.ml#L220
(`Load_path.get_paths ()` returns the list of entries in reverse order).
- In `Cmt_format.cmt_loadpath`, the list of entries in the load path is being recorded in reverse order: https://github.com/ocaml/ocaml/blob/e68b75eb823c27894c5a2fcf30f1166029abffb1/file_formats/cmt_format.ml#L184
(see eg https://github.com/ocaml/ocaml/blob/e68b75eb823c27894c5a2fcf30f1166029abffb1/tools/read_cmt.ml#L178)
- PPX contexts are storing the load path in reverse order, and using it when restored as if it were in normal order: https://github.com/ocaml/ocaml/blob/e68b75eb823c27894c5a2fcf30f1166029abffb1/parsing/ast_mapper.ml#L822 and https://github.com/ocaml/ocaml/blob/e68b75eb823c27894c5a2fcf30f1166029abffb1/parsing/ast_mapper.ml#L892
(because `Load_path.init` expects the list in the normal order but `Load_path.get_paths ()` returns it reversed, as mentioned above).

### Description of the commits

- To keep this from happening in the future, the first commit introduces a type of "reversed lists" or "lists with an efficient cons at the end" that need to be explicitly converted into "normal lists". The idea is to use this type instead of sprinkling `List.rev`'s everywhere. I also plan to use this in #9056.

- The second commit uses this type in the implementation of `Load_path` and applies the necessary explicit reversal to return entries in the normal order in `Load_path.get_paths`.

- Finally, the last commit (which is optional) applies the new type to `Clflags.include_dirs`. I didn't uncover any reversal bugs when doing this, but may be useful for future.

nojb committed:
Add a specific type for "reversed lists"

nojb committed:
Fix ordering of directories in load path

nojb commented on typing/typemod.ml:
```diff
@@ -164,7 +164,7 @@ let initial_env ~loc ~safe_string ~initially_opened_module
   in
   let units =
-    List.rev_map Env.persistent_structures_of_dir (Load_path.get ())
+    List.map Env.persistent_structures_of_dir (Load_path.get ())
```
This is the only place I could find where the caller was correctly reversing the result of `Load_path.get` before using it. After the changes in the PR, the results are being returned in normal order, so this reversal is no longer needed.

gasche commented:
> To keep in mind when reading the code: earlier entries in the load path take precedence over later entries (ie written later in the command line).

I am surprised to learn this, is it documented somewhere? Naively I would expect `-I foo -I bar` to have a semantics resembling having

```
open Foo
open Bar
```

in the source program, for "virtual" modules `Foo` and `Bar` containing all the compilation units in their respective directory, with later `-I` modules shadowing earlier ones.

(In practice I expect that the question rarely occurs in OCaml due to the impossibility to work correctly with two compilation units of the same name in the load path.)

Partly self-replying to myself: the manpage for `ocamlc` says

> Directories added with -I are searched after the current directory, in the order in which they were given on the  command line, but before the standard library directory.

which I guess serves as documentation of this surprising fact. Is there a good reason for using this order instead of the (possibly more natural) other one with a shadowing semantics?

nojb commented:
> > To keep in mind when reading the code: earlier entries in the load path take precedence over later entries (ie written later in the command line).
> 
> I am surprised to learn this, is it documented somewhere?

You are not the only one :)

It is documented in the manual (as you found out).

> which I guess serves as documentation of this surprising fact. Is there a good reason for using this order instead of the (possibly more natural) other one with a shadowing semantics?

Not sure. It matches `gcc` and the like, for what that's worth.

gasche commented on utils/misc.mli:
```diff
@@ -682,3 +682,25 @@ module Magic_number : sig
+
+  val empty : 'a t
+  val to_list : 'a t -> 'a list
+  val to_list_append : 'a t -> 'a list -> 'a list
```
on callsites I would find `append_to_list` more natural, and actually I think that `rev_append` would be fine.

gasche commented on utils/misc.mli:
```diff
@@ -682,3 +682,25 @@ module Magic_number : sig
+  val filter_map : ('a -> 'b option) -> 'a t -> 'b t
+
+  val add_at_end : 'a -> 'a t -> 'a t
+  (** Add an element at the {e end} of the list. *)
```
I am used to the name `snoc` (cons in reverse) for this operation.

gasche reviewed (commented):
Two remarks about this change:
1. The documentation says something and then the implementation mostly does something else. I think it would be worth making sure that users do not depend on what the implementation does, rather than what the documentation says. (Usually when they are contradictory, people test and assume that this is the intended behavior.) It may be the case that the only case where the order is handled correctly is the one that matter (construction of the initial typing environment), or that in fact nobody has unit-name conflicts in their load path so we never observe the difference.

2. In principle I like the idea of a type to denote the fact that a list is in the unexpected order, but another think that I like more is *not* having lists in the unexpected order. RevList sounds like a nice thing to use *locally* while building the list in reverse, before putting it back in the unsurprising order. From a first skim at your PR, I have the impression that instead of you are using it to program safely in a situation where everyone sees the wrong-order list. Could we consider instead ensuring that, except in the command-flag-handling logic, consumers see the list in the "right" order? It might even be the case that if we do this, most use-cases for `RevList` vanish (in this patch) and we can do fine with just some hungarian notation (using `rev_foo` for the variables accumulating in reverse order).

nojb commented:
> 2. In principle I like the idea of a type to denote the fact that a list is in the unexpected order, but another think that I like more is _not_ having lists in the unexpected order. RevList sounds like a nice thing to use _locally_ while building the list in reverse, before putting it back in the unsurprising order. From a first skim at your PR, I have the impression that instead of you are using it to program safely in a situation where everyone sees the wrong-order list. Could we consider instead ensuring that, except in the command-flag-handling logic, consumers see the list in the "right" order? It might even be the case that if we do this, most use-cases for `RevList` vanish (in this patch) and we can do fine with just some hungarian notation (using `rev_foo` for the variables accumulating in reverse order).

Unless I am misremembering `RevList` is only visible via the command-flag-handling logic. This is the third commit and is optional, as it does not fix any bug. In `Load_path` the use of `RevList` is hidden in the implementation.

dbuenzli commented:
> I am surprised to learn this,

FWIW I'm not surprised. I think you have the wrong take on it: you need to understand it as a search procedure. It corresponds exactly to the way  `PATH`-like variables are searched. That is  gather all the `-I` includes into a list and then search from left to right in it.

nojb commented:
Just to keep in mind: the `string list` in `val Clflags.include_dirs : string list ref` will probably have to go away for something a bit more elaborate in #9056 anyway, so the third commit here will probably have  another shot at coming back under the hood in that PR. So no problem for me to remove it from this one.

jeremiedimino commented on driver/compmisc.ml:
```diff
@@ -23,16 +25,22 @@ open Compenv
+      [ !first_include_dirs;
+        RevList.of_list Config.flexdll_dirs;
+        dirs;
+        !last_include_dirs ]
```
The new version is definitely easier to reason about :)

ghost commented:
@nojb, in the new version I'm not entirely sure `SMap.find fn !files` and `Misc.find_in_path (get_paths ()) fn` do the same thing in `Load_path.find`. More precisely, given that we shadow previous entries in the map in `Load_path.add`, it seems to me that going through the cache implements the search from right to left while going through `Misc.find_in_path` implements the search from left to right. Do you agree?

I was also wondering if we shouldn't introduce an abstract type for dealing with load paths. The `Load_path` module is for the "main load path"  + a cache, but we could reorganise the code to have a generic module for dealing with load paths in general, and then another one for dealing with the main one and adding a cache for fast repeated lookups. At least, we wouldn't have to think about the order of the list and we could use the abstract type uniformly instead of `string list`.

nojb commented:
> @nojb, in the new version I'm not entirely sure `SMap.find fn !files` and `Misc.find_in_path (get_paths ()) fn` do the same thing in `Load_path.find`. More precisely, given that we shadow previous entries in the map in `Load_path.add`, it seems to me that going through the cache implements the search from right to left while going through `Misc.find_in_path` implements the search from left to right. Do you agree?

You are right, in some cases.

- When calling `init`, the argument is in the normal, left-to-right, order, and `add` is iterated on `List.rev (argument)`, so that in fact entries are inserted in the map in right-to-left order, ie left takes precedence over right, which matches up with the behaviour of `find_in_path`.

- On the other hand, if one calls `Load_path.add` directly, not via `Load_path.init` (eg when using `#directory` in the toplevel) indeed, the argument will take precedence over the previously added directories in the cache, and it won't match with the behaviour of `find_in_path`. One solution is to "refresh" the whole load path each time it is modified (like it is done in `Load_path.remove_dir`)...

> I was also wondering if we shouldn't introduce an abstract type for dealing with load paths. The `Load_path` module is for the "main load path" + a cache, but we could reorganise the code to have a generic module for dealing with load paths in general, and then another one for dealing with the main one and adding a cache for fast repeated lookups. At least, we wouldn't have to think about the order of the list and we could use the abstract type uniformly instead of `string list`.

Absolutely! This is how I am approaching #9056 (the third commit is a disguised version of that, actually). But I guess probably better to leave it out of this PR...

nojb committed:
Rename RevList.{append_at_end => snoc}

nojb committed:
Add RevList.rev_iter

nojb committed:
Refresh load path when adding a new entry

nojb commented:
In order to focus the discussion, I removed the commit about `Clflags.include_dirs`, and added a commit fixing the problem pointed out by @jeremiedimino (refreshing the load path when calling `Load_path.add{,_dir}`).

ghost commented:
@nojb, hmm it's more the first `add` function that I'm concerned about :thinking:; We add a the directory at the end of the `dirs` list, but we shadow entries in the `files` map. What about not adding a file in the map if it's already there? Then both code path should do the same thing and we don't need to reset and re-add everything when we add a directory.

nojb committed:
Only add an entry to the cache if not already present in order to enforce left-to-right precedence

nojb committed:
Revert use of RevList

nojb commented:
> @nojb, hmm it's more the first `add` function that I'm concerned about 🤔; We add a the directory at the end of the `dirs` list, but we shadow entries in the `files` map.

Well, the first `add` function is now only used in `init` and `remove_dir` which means that we are starting from an empty cache and adding the entries in reverse order, so it is OK.

> What about not adding a file in the map if it's already there? Then both code path should do the same thing and we don't need to reset and re-add everything when we add a directory.

Good idea. I pushed a commit with this suggestions + some comments that should help reading the code.

Finally, I pushed a commit removing the use of `RevList`. I think it is a good idea to have a dedicated type, but not worth polluting this PR with it, as it was being used only in one or two places.

nojb committed:
Fix

nojb commented:
> > @nojb, hmm it's more the first `add` function that I'm concerned about 🤔; We add a the directory at the end of the `dirs` list, but we shadow entries in the `files` map.
> 
> Well, the first `add` function is now only used in `init` and `remove_dir` which means that we are starting from an empty cache and adding the entries in reverse order, so it is OK.
> 

I hadn't fully understood your comment about the `dirs` variable; I pushed one more commit with a fix for this mismatch (so that the `dirs` variable is always kept in reverse order).

ghost commented:
Ack, having another read now

ghost commented:
I had another look and the code of `load_path.ml` looks correct to me, thanks for reworking it and adding comments!

nojb committed:
Changes

nojb commented:
> LGTM!

Thanks for the review!

@gasche are you happy with the current patch?

gasche commented:
I haven't had time to look at it, sorry. Please discard my earlier cursory comments, and feel free to go ahead and merge.

nojb commented:
> 1. The documentation says something and then the implementation mostly does something else. I think it would be worth making sure that users do not depend on what the implementation does, rather than what the documentation says. (Usually when they are contradictory, people test and assume that this is the intended behavior.) It may be the case that the only case where the order is handled correctly is the one that matter (construction of the initial typing environment), or that in fact nobody has unit-name conflicts in their load path so we never observe the difference.

In fact I just found a bug report about the inverse order of paths when passed to the linker (fixed by this PR): #7716
